src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DoNotAddFinalizerToMemoryManagerTDerivedTypes.cs

```diff
@@ -0,0 +1,55 @@
+
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Analyzer.Utilities.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.NetCore.Analyzers.Runtime
+{
+    /// <summary>
+    /// CA2014: Do not add finilazer for MemoryManaged&lt;T&gt; derived types.
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public sealed class DoNotAddFinalizerToMemoryManagerTDerivedTypes : DiagnosticAnalyzer
+    {
+        internal const string RuleId = "CA2014";
+
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.DoNotAddFinalizerToMemoryManagedTDerivedTypesTitle), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.DoNotAddFinalizerToMemoryManagedTDerivedTypesMessage), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(MicrosoftNetCoreAnalyzersResources.DoNotAddFinalizerToMemoryManagedTDerivedTypesDescription), MicrosoftNetCoreAnalyzersResources.ResourceManager, typeof(MicrosoftNetCoreAnalyzersResources));
+
+        internal static readonly DiagnosticDescriptor FinilizerRule = DiagnosticDescriptorHelper.Create(RuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.Reliability,
+            RuleLevel.BuildWarning,
+            description: s_localizableDescription,
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(FinilizerRule);
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
+            analysisContext.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            analysisContext.RegisterSymbolAction(AnalyzeSymbol, SymbolKind.NamedType);
+        }
+
+        private static void AnalyzeSymbol(SymbolAnalysisContext context)
+        {
+            var namedTypeSymbol = (INamedTypeSymbol)context.Symbol;
+            var memoryManager1 = context.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemBuffersMemoryManager1);
+
+            if (memoryManager1 != null && namedTypeSymbol.DerivesFromOrImplementsAnyConstructionOf(memoryManager1))
+            {
+                if (namedTypeSymbol.HasFinalizer())
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(FinilizerRule, namedTypeSymbol.Locations[0], namedTypeSymbol.Name));
```

Comment:
It is pretty simple rule, but drawback is because it is registered for the class name the warning line number is wrong, I mean not the line of the distractor, not sure if that need to be exact line number, if that is the case rule need to be changed

Reply:
Instead of `namedTypeSymbol.HasFinalizer()` you could do

```C#
var finalizerMethod = namedTypeSymbol.GetMembers().FirstOrDefault(m => m is IMethodSymbol method && method.IsFinalizer());

if (finalizerMethod != null)
{
    report a diagnostic against finalizerMethod
}
```

That way you still only run once per type.  (It's really just expanding the HasFinalizer method to return the result instead of a boolean.)